### ISSUE TITLE
[Linter] <Double force casting>

### ISSUE THREAD

**Issue by jacob-tucker** [onflow/cadence-tools#217]
### Current Behavior

Right now if you double force cast, for example, `test["testing"]!!`, the linter will tell you that you are performing an `unecessary force unwrap`.

However, given the fact that it's possible for dictionaries to have `nil` values, this actually may be needed some times to convert, for example, a `String??` to a `String`

### Expected Behavior

The linter should not complain about a double force cast if the type is a double optional

### Steps To Reproduce

```cadence
pub fun main(): AnyStruct {
  let bin: {String: String} = { "gg": "10" }
  var v : {String: AnyStruct} =  { "deniz": bin["gg"] }
  let thing = v["deniz"]!!.getType()
  return thing
}
```

### Environment

```markdown
- Cadence version: 0.40.0
- Network: Emulator (or any)
```

**Comment by turbolent** [comment]
Doesn't this work as expected: 
- The value type of `v` is `AnyStruct`
- `v["deniz"]` results in `AnyStruct?`
- Two force-unwraps are used, but only one is necessary to unwrap

**Comment by bluesign** [comment]
I think problem is: `AnyStruct` covering optional types. This is working as intended ( warning ) for sure. 

But I think this should be only allowed to written as :

```cadence 
pub fun main(): AnyStruct {
  let bin: {String: String} = { "gg": "10" }
  var v : {String: AnyStruct?} =  { "deniz": bin["gg"] }
  let thing = v["deniz"]!!.getType()
  return thing
}
```

But as `AnyStruct` covers `AnyStruct?` this is going into area of cannot be statically determined.

**Comment by turbolent** [comment]
Yeah, that's what I tried to point out: The analysis works only on static types, and from that perspective the implementation is correct.

However, you're right that `AnyStruct` as the top-type covers any other dynamic type, so the dynamic type of the underlying value _could_ actually be optional. This is unlike any other type (e.g. think through the example with the dictionary's value type being e.g. `Int`).

So the analysis could be improved here to ignore `Any`/`AnyStruct`/`AnyResouce`, to avoid reporting a warning that the operator is unnecessary, where it might actually be necessary. 👍

**Comment by turbolent** [comment]
In particular, the exception should be added here: https://github.com/onflow/cadence-tools/blob/55644ee1060f24eea572d955d657d501c311b3bc/lint/unnecessary_force_analyzer.go#L60

**Comment by bluesign** [comment]
> However, you're right that `AnyStruct` as the top-type covers any other dynamic type, so the dynamic type of the underlying value _could_ actually be optional. This is unlike any other type (e.g. think through the example with the dictionary's value type being e.g. `Int`).

Btw is it possible to change hierarchy, like putting `Optional` higher than `AnyStruct`? This maybe totally stupid idea. I have no idea how other languages work, or theory of this stuff as I don't have any CS background.

**Comment by turbolent** [comment]
> is it possible to change hierarchy, like putting `Optional` higher than `AnyStruct`?

No, `AnyStruct` is the top type (for values / non-resources). A good analogy is thinking of optionals as "zero-or-one containers". Moving them up in the hierarchy makes not much sense, just like moving up arrays ("zero-or-more containers") makes not much sense.

**Comment by bluesign** [comment]
In swift this one errors out for example.

```swift
var d : Int? = 1
var myDict : [String : AnyObject] = [ "deniz" : d]
```
`cannot convert value of type 'Int?' to expected dictionary value type 'AnyObject'`

Is it because Optional is not somehow AnyObject there?

**Comment by turbolent** [comment]
These are great questions, and am happy to answer, but let's move such questions to the forum, they are unrelated to the issue.
